### PR TITLE
Make taxRate part of item details for all tax items added to the invoice

### PR DESCRIPTION
**Pull request by vlaskhilkevich**
Using Invoice's `itemDetails` field to preserve for TAX items effective tax rate obtained from vertex as json object in the format (example):
`{"taxRate":0.09975}`

This field can be used for more information if needed in the future

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java [hunk 1]
```diff
@@ -305,6 +313,56 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
         }
     }
 
+    private InvoiceItem createTaxInvoiceItem(final InvoiceItem taxableItem, final UUID invoiceId, @Nullable final InvoiceItem adjustmentItem, final BigDecimal calculatedTax, @Nullable final String description, @Nullable Double taxRate) {
+        final InvoiceItem taxItem = buildTaxItem(taxableItem, invoiceId, adjustmentItem, calculatedTax, description);
+        if (taxItem == null || taxRate == null) {
+            return taxItem;
+        }
```
Comment (singh-sukhvir) [inline src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:320]: should we return null in above case. Also, shouldn't we also check if taxrate = 0, we may want to skip creating taxitem in that case as well.

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java [hunk 2]
```diff
@@ -305,6 +313,56 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
...
+                                             .validate().build());
+    }
+
+    private String createTaxItemDetails(@Nonnull final Double effectiveRate) {
+        try {
+            return objectMapper.writeValueAsString(ImmutableMap.of("taxRate", String.valueOf(effectiveRate)));
+        } catch (JsonProcessingException ignored) {
+            return null;
+        }
+    }
```
Comment (singh-sukhvir) [inline src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:364]: can we make this generic in the sense  like `private String appendItemDetails(key, value)` and let caller decide key as well.

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java [hunk 3]
```diff
@@ -305,6 +313,56 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
...
+                                             .withUpdatedDate(taxItem.getUpdatedDate())
+                                             .validate().build());
+    }
+
+    private String createTaxItemDetails(@Nonnull final Double effectiveRate) {
+        try {
+            return objectMapper.writeValueAsString(ImmutableMap.of("taxRate", String.valueOf(effectiveRate)));
+        } catch (JsonProcessingException ignored) {
+            return null;
+        }
```
Comment (singh-sukhvir) [inline src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:363]: also, we whouldn't have a tax-item without eefective tax rate info. If we got null or 0 values from vertex we shouldn't create the tax item at all as I mentioned in my other comment.

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 4]
```diff
@@ -231,6 +231,40 @@ public void testTaxDescription() throws Exception {
         assertEquals("CA STATE TAX", result.get(0).getDescription());
     }
 
+    @Test(groups = "fast")
+    public void testTaxEffectiveRate() throws Exception {
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:237]: can we break into 3 test cases - , no tax rate returned, 0 tax rate returned , non-zero tax rate returned

**Review by singh-sukhvir — CHANGES_REQUESTED** [review]
Left some comments to check some preconditions.

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 1 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:320]: Creating/not creating of the taxItem base on existence of the effective rate will change existing behaviour and I am not sure if this is correct to do so, I wanted to leave it as is.

Also I don't think we need to check taxRate on 0 and control tax item creation as it is already controlled by tax amount in the existing code

Returning taxItem here is returning basic (not derived taxItem)
```
 final InvoiceItem taxItem = createTaxInvoiceItem(taxableItem, invoiceId, adjustmentItem, calculatedTax, description, transactionLineDetailModel.getEffectiveRate());
                if (taxItem != null) {
                    invoiceItems.add(taxItem);
                }
```

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 2 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:364]: I didn't get append thing, append to what?

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 3 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:363]: I wanted to not to change existing behaviour, imo it's safe implementation to not miss tax if effectiveTaxRate is not presented by vertex by some reason. We return not derived basic taxItem created by existing method.

```
final String taxItemDetails = createTaxItemDetails(taxRate);
        if (taxItemDetails == null) {
            return taxItem;
        }
```

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 4 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:237]: I added few test cases and the one showing the behaviour when the amount is 0 (usually this is what happens when taxRate is 0). This should be regular scenario - https://github.com/killbill/killbill-vertex-plugin/blob/a1a9e5f6ece10d3aa332a3ffbc2bbc1831da8357/src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java#L280, but this one is an exception - https://github.com/killbill/killbill-vertex-plugin/blob/a1a9e5f6ece10d3aa332a3ffbc2bbc1831da8357/src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java#L297 and

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 2 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java]: updated to be generic by passing the map with the details into the method

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 1 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:320]: And the basic taxItem will be null in case of not positive amount:

```
 if (amount != null && BigDecimal.ZERO.compareTo(amount) != 0) {
            return adjustmentItem != null ? PluginInvoiceItem.createTaxItem(originalItem, invoiceId, adjustmentItem.getStartDate(), adjustmentItem.getEndDate(), amount, (String)MoreObjects.firstNonNull(description, "Tax")) : PluginInvoiceItem.createTaxItem(originalItem, invoiceId, amount, (String)MoreObjects.firstNonNull(description, "Tax"));
        } else {
            return null;
        }
```

Comment (singh-sukhvir) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:320]: based on above, tax rate should never be null for `taxItem` as the amount is non-zero. So, we don't even need that additional` || taxRate == null` check

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 5]
```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
...
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(0, result.size());
+    }
+
+    @Test(groups = "fast")
+    public void testComputeWhenEffectiveTaxRateZeroButAmountIsNot() throws Exception {
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:299]: this is not a valid usecase

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 6]
```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
...
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(1, result.size());
+        assertNull(result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+    }
+
+    @Test(groups = "fast")
+    public void testComputeWhenEffectiveTaxRateZero() throws Exception {
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:280]: this test is more of "When tax is zero" instead of `WhenEffectiveTaxRateZero` because we have no condition on 0 tax-rate

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 7]
```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
...
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
+        given(taxResponse.getData()).willReturn(apiResponseData);
+        final TaxesType taxesType = new TaxesType();
+        taxesType.setCalculatedTax(MOCK_TAX_AMOUNT_1_01);
+        given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
+
+        //given tax effective rate is not present
+        taxesType.setEffectiveRate(null);
+
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:271]: in this case, we are setting amount != 0 but tax rate as null which is not a practical usecase

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 8]
```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
...
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals("{\"taxRate\":0.09975}", result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+
+        //given effective rate is not present
+        taxesType.setEffectiveRate(null);
+
+        //then no item details persisted
+        result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertNull(result.get(0).getItemDetails());
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:256]: it is possible that taxRate is not 0 and amount is 0 but vice versa cannot be true.

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 9]
```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
...
+        taxesType.setEffectiveRate(0d);
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(1, result.size());
+        assertEquals("{\"taxRate\":0.0}", result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+    }
+
+    private void checkTaxItemFields(final InvoiceItem taxItem) {
```
Comment (singh-sukhvir) [inline src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:315]: To my understanding, below are the practical cases -

1. InvoiceItem has 0 tax - No tax item should be generated
2. InvoiceItem has non-zero tax - taxItem with ItemDetails is generated.

Other scenario where tax-rate is null should not happen if tax-amount != 0

**Review by singh-sukhvir — APPROVED** [review]
Looks good mostly. Tests need some work. Some tests look redundant as well

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 1 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java:320]: taxRate can be theoretically be null (and we don't control if at some point there will be a bug on vertex side), hence I checked on that to not to construct derived tax item without root information why it is being created (taxRate). Otherwise we will return taxDetails = {"taxRate": null} or same taxItem but created as duplicate (with taxRate as null value if there will be an exception during serialization). I don't see the problem to check on null and make life easier here.

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 5 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:299]: this usecase not valid from business perspective and common sense, but i described my thoughts on it [the reply](https://github.com/killbill/killbill-vertex-plugin/pull/42#discussion_r1690516325). 

I created these test case for these 2 reasons:
1. It shows how system behaves in this edge case
2. It shows that we still create tax items even if taxRate is 0 (tests that we rely not on taxRate but on the amount)

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 9 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:315]: I think the key word is "should not happen". But it can, doesn't it?

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 6 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:278]: removed not needed tests

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 9 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java:291]: discussed

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 8 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: discussed

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 7 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: discussed

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 5 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: discussed

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java (hunk 4 above)
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: done

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java [hunk 10]
```diff
@@ -315,8 +316,12 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
...
         final InvoiceItem taxItem = buildTaxItem(taxableItem, invoiceId, adjustmentItem, calculatedTax, description);
-        if (taxItem == null || taxRate == null) {
-            return taxItem;
+        if (taxItem == null) {
+            return null;
+        }
+
+        if (taxRate == null) {
+            taxRate = calculatedTax.divide(taxableItem.getAmount(), 5, RoundingMode.HALF_UP).doubleValue();
         }
```
[the remark this answers is not in the document]
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java]: make sense, fixed

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 11]
```diff
@@ -302,10 +279,9 @@ public void testComputeWhenEffectiveTaxRateZeroButAmountIsNot() throws Exception
         taxesType.setCalculatedTax(MOCK_TAX_AMOUNT_1_01);
         given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
 
-        //given tax effective rate is present
         taxesType.setEffectiveRate(0d);
 
-        //then it persisted in item details invoice item field
+        //tax item with tax rate in item details is added to the result
```
[the remark this answers is not in the document]
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: i think it worth to have separate test to cover the above as we show and check that we control tax creation not by tax rate

File: src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java [hunk 12]
```diff
@@ -125,10 +136,16 @@ public void setUp() throws Exception {
...
 
         given(vertexDao.getSuccessfulResponses(any(UUID.class), any(UUID.class))).willReturn(Collections.emptyList());
-        given(vertexApiClient.calculateTaxes(any(SaleRequestType.class))).willReturn(taxResponse);
+
+        given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
+        given(taxesType.getCalculatedTax()).willReturn(MOCK_TAX_AMOUNT_1_01);
 
         given(responseLineItem.getLineItemId()).willReturn(TAX_ITEM_ID.toString());
         given(responseLineItem.getTotalTax()).willReturn(MOCK_TAX_AMOUNT_1_01);
+        given(responseLineItem.getTotalTax()).willReturn(MOCK_TAX_AMOUNT_1_01);
```
[the remark this answers is not in the document]
Comment (vlaskhilkevich) [reply src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java]: fixed, thanks

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java [hunk 13]
```diff
@@ -321,6 +321,7 @@ private InvoiceItem createTaxInvoiceItem(final InvoiceItem taxableItem, final UU
...
 
         if (taxRate == null) {
+            logger.info("The tax rate is not provided in the Vertex response for the tax item with ID: {} and calculated tax: {}", taxItem.getId(), calculatedTax);
```
[the remark this answers is not in the document]
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java]: done

File: src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java (hunk 3 above)
Comment (vlaskhilkevich) [reply src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java]: discussed

**Review by singh-sukhvir — APPROVED** [review]
looks good
